### ISSUE TITLE
Runtime crash: double free or corruption at startup

### ISSUE THREAD

**Issue by hfc2x** [dxx-rebirth/dxx-rebirth#653]
#### Operating System Environment

* Microsoft Windows
  * [ ] Microsoft Windows XP
  * [ ] Microsoft Windows Vista
  * [ ] Microsoft Windows 7
  * [ ] Microsoft Windows 8
  * [ ] Microsoft Windows 8.1
  * [ ] Microsoft Windows 10

* Mac OS X
  * [ ] Mac OS X 10.10
  * [ ] Mac OS X 10.11
  * [ ] Mac OS X 10.12
  * [ ] Mac OS X 10.13
  * [ ] Mac OS X 10.14
  * [ ] Mac OS X 10.15

* Debian
  * [ ] Debian Stretch
  * [ ] Debian Buster
  * [ ] Debian Bullseye
  * [ ] Debian Sid
* Fedora
  * [ ] Fedora 28
  * [ ] Fedora 29
  * [ ] Fedora 30
  * [ ] Fedora 31
  * [ ] Fedora 32
  * [ ] Rawhide
* Ubuntu
  * [ ] Ubuntu 16.04 LTS (Xenial Xerus)
  * [ ] Ubuntu 18.04 LTS (Bionic Beaver)
  * [ ] Ubuntu 18.10 (Cosmic Cuttlefish)
  * [ ] Ubuntu 19.04 (Disco Dingo)
  * [ ] Ubuntu 19.10 (Eoan Ermine)
  * [ ] Ubuntu 20.04 LTS (Focal Fossa)

* [x] Arch

* [ ] Gentoo

* [ ] OpenBSD
* [ ] FreeBSD
* [ ] NetBSD

#### CPU environment

* [ ] x86 (32-bit Intel/AMD)
* [x] x86\_64 (64-bit Intel/AMD)
* [ ] ARM (32-bit)
* [ ] ARM64 (64-bit; sometimes called AArch64)

### Description

Descent 1 crashes on startup on Manjaro Linux (5.15.59-1 kernel) with an error message stating "double free or corruption (fasttop)". The game was built using the current commit (1c8ff1ac0f137aef68a50611c6714cd1cdf0a043), by using the options `scons d1x=1 sdl2=yes`. Compiled with SDL2, because SDL1.2 doesn't work. Installed all the required dependencies listed in INSTALL.txt. All the game data is in the same folder as the d1x-rebirth executable.

Trying to launch the game starts with a full screen, that immediately closes. Terminal output is as follows:
`./d1x-rebirth -hogdir .`
```
Type './d1x-rebirth -help' for a list of command-line options.

D1X-Rebirth v0.61 0.60.0-beta2-1862-g1c8ff1ac0f13  Aug 10 2022 13:16:51
This is a MODIFIED version of Descent, based on Registered v1.5 Jan 5, 1996.
Copyright (C) 1994, 1995 Parallax Software Corporation
DESCENT is a trademark of Interplay Productions, Inc.
Copyright (C) 2005-2013 Christian Beckhaeuser, 2013-2017 Kp
Using SDL_mixer library v2.6.1
sdl-joystick: 0 joysticks
DXX-Rebirth: OpenGL: disabling automatic GL sync since VSync is turned off
double free or corruption (fasttop)
zsh: IOT instruction (core dumped)  ./d1x-rebirth -hogdir .
```

### Steps to Reproduce
Just launch the game through the terminal by typing `./d1x-rebirth -hogdir .`. Fullscreen starts, and then immediately closes with error mentioned above.

* Engines affected:
  * [x] D1X-Rebirth
  * [ ] D2X-Rebirth

* Game modes affected:
  * [ ] Single player
  * [ ] Multiplayer competitive
    * [ ] Anarchy
    * [ ] Team Anarchy
    * [ ] Robo-Anarchy
    * [ ] Capture the Flag
    * [ ] Bounty
  * [ ] Multiplayer cooperative

* Players affected by issue:
  * [ ] Game host
  * [ ] Game guests

**Comment by vLKp** [comment]
I may miss some bugs, but a crash at startup is hard to miss.  This is not reproducible here, using the same commit and SDL choice as shown in the initial report.  Generally, if a core dump is created, it is useful to post the backtrace from the core dump.  Please provide it, so that I can see specifically where the double free is detected.

**Comment by mmetalfan** [comment]
I would also like to add that I am having similar issues starting the game except I get a general segmentation fault. 
./d2x-rebirth -hogdir .                                                                                                                                                                                                                                                                  ✔ 

Type ./d2x-rebirth -help' for a list of command-line options.

D2X-Rebirth v0.58.1  Jul 15 2016 14:56:23
This is a MODIFIED version of Descent 2, based on Full Version v1.2.
Copyright (C) 1994-1996 Parallax Software Corporation
DESCENT is a trademark of Interplay Productions, Inc.
Copyright (C) 1999 Peter Hawkins, 2002 Bradley Bell, 2005-2011 Christian Beckhaeuser

Using SDL_mixer library
sdl-joystick: found 0 joysticks
zsh: segmentation fault (core dumped)  ./d2x-rebirth -hogdir .

**Comment by hfc2x** [comment]
> Generally, if a core dump is created, it is useful to post the backtrace from the core dump. Please provide it, so that I can see specifically where the double free is detected.

How do I provide that? I have the output of gamelog.txt here, by launching the game with `-debug -verbose -safelog`, but I'm not sure this is what you're requesting:
```
18:23:53.965814 cmd_addcommand: added alias
18:23:53.965860 cmd_addcommand: added unalias
18:23:53.965876 cmd_addcommand: added echo
18:23:53.965890 cmd_addcommand: added exec
18:23:53.965905 cmd_addcommand: added help
18:23:53.965919 cmd_addcommand: added wait
18:23:53.965934 cmd_addcommand: added set
18:23:53.965957 PHYSFS: Listing contents of Search Path.
18:23:53.965974 PHYSFS: [/home/redacted/.d1x-rebirth/] is in the Search Path.
18:23:53.965990 PHYSFS: [.] is in the Search Path.
18:23:53.966005 PHYSFS: [/home/redacted/.d1x-rebirth/data] is in the Search Path.
18:23:53.966089 PHYSFS: * We've got [.directory].
18:23:53.966106 PHYSFS: * We've got [PLAYER.PLR].
18:23:53.966123 PHYSFS: * We've got [chaos.hog].
18:23:53.966137 PHYSFS: * We've got [chaos.msn].
18:23:53.966151 PHYSFS: * We've got [d1x-rebirth].
18:23:53.966166 PHYSFS: * We've got [data].
18:23:53.966181 PHYSFS: * We've got [descent.b50].
18:23:53.966195 PHYSFS: * We've got [descent.bat].
18:23:53.966209 PHYSFS: * We've got [descent.cfg].
18:23:53.966224 PHYSFS: * We've got [descent.dem].
18:23:53.966238 PHYSFS: * We've got [descent.faq].
18:23:53.966253 PHYSFS: * We've got [descent.hog].
18:23:53.966268 PHYSFS: * We've got [descent.ico].
18:23:53.966282 PHYSFS: * We've got [descent.m50].
18:23:53.966297 PHYSFS: * We've got [descent.phx].
18:23:53.966312 PHYSFS: * We've got [descent.pig].
18:23:53.966327 PHYSFS: * We've got [descent2.adv].
18:23:53.966341 PHYSFS: * We've got [descentg.ini].
18:23:53.966356 PHYSFS: * We've got [descentr.exe].
18:23:53.966371 PHYSFS: * We've got [devteam.pcx].
18:23:53.966386 PHYSFS: * We've got [gamelog.txt].
18:23:53.966401 PHYSFS: * We've got [hmidet.386].
18:23:53.966415 PHYSFS: * We've got [hmidrv.386].
18:23:53.966430 PHYSFS: * We've got [hmimdrv.386].
18:23:53.966445 PHYSFS: * We've got [level18.dem].
18:23:53.966459 PHYSFS: * We've got [miniboss.dem].
18:23:53.966474 PHYSFS: * We've got [pcxview.exe].
18:23:53.966489 PHYSFS: * We've got [phqghume.ayl].
18:23:53.966504 PHYSFS: * We've got [readme.txt].
18:23:53.966518 PHYSFS: * We've got [sdl.dll].
18:23:53.966533 PHYSFS: * We've got [sdl_net.dll].
18:23:53.966548 PHYSFS: * We've got [setup.exe].
18:23:53.966563 PHYSFS: Checking supported archive types.
18:23:53.966579 PHYSFS: Supported archive: [ZIP], which is [PkZip/WinZip/Info-Zip compatible].
18:23:53.966594 PHYSFS: Supported archive: [7Z], which is [7zip archives].
18:23:53.966609 PHYSFS: Supported archive: [GRP], which is [Build engine Groupfile format].
18:23:53.966624 PHYSFS: Supported archive: [PAK], which is [Quake I/II format].
18:23:53.966639 PHYSFS: Supported archive: [HOG], which is [Descent I/II HOG file format].
18:23:53.966654 PHYSFS: Supported archive: [MVL], which is [Descent II Movielib format].
18:23:53.966669 PHYSFS: Supported archive: [WAD], which is [DOOM engine format].
18:23:53.966684 PHYSFS: Supported archive: [SLB], which is [I-War / Independence War Slab file].
18:23:53.966699 PHYSFS: Supported archive: [ISO], which is [ISO9660 image file].
18:23:53.966714 PHYSFS: Supported archive: [VDF], which is [Gothic I/II engine format].
18:23:53.967284 PHYSFS: append canonical directory "./descent.hog" to search path from relative name "descent.hog"
18:23:53.967521 D1X-Rebirth v0.61 0.60.0-beta2-1862-g1c8ff1ac0f13  Aug 10 2022 13:16:51
18:23:53.967538 This is a MODIFIED version of Descent, based on Registered v1.5 Jan 5, 1996.
18:23:53.967552 Copyright (C) 1994, 1995 Parallax Software Corporation
18:23:53.967567 DESCENT is a trademark of Interplay Productions, Inc.
18:23:53.967581 Copyright (C) 2005-2013 Christian Beckhaeuser, 2013-2017 Kp
18:23:53.967597 D1X-Rebirth built with PhysFS 3.0.2; loaded with PhysFS 3.0.2
18:23:53.967628 D1X-Rebirth built with libSDL 2.0.22; loaded with libSDL 2.0.22
18:23:53.967647 D1X-Rebirth built with SDL_image 2.6.1; loaded with SDL_image 2.6.1
18:23:53.967672 D1X-Rebirth built with SDL_mixer 2.6.1; loaded with SDL_mixer 2.6.1
18:23:53.967692 D1X-Rebirth built with libpng version 1.6.37; loaded with libpng version 1.6.37
18:23:53.967707 Getting settings from DESCENT.CFG...
18:23:53.967840 PHYSFS: Adding archives to the game.
18:23:53.987208 Using SDL_mixer library v2.6.1
18:23:54.022929 sdl-joystick: 0 joysticks
18:23:54.217611 Going into graphics mode...
18:23:54.217861 DXX-Rebirth: OpenGL: version 4.6 (4.6 (Compatibility Profile) Mesa 22.1.4)
18:23:54.217912 DXX-Rebirth: OpenGL: GL_EXT_texture_filter_anisotropic available, max anisotropy: 16.000000
18:23:54.217966 DXX-Rebirth: OpenGL: GL_ARB_sync available
18:23:54.217998 DXX-Rebirth: OpenGL: vendor: Intel
18:23:54.218028 DXX-Rebirth: OpenGL: renderer: Mesa Intel(R) HD Graphics 630 (KBL GT2)
18:23:54.218057 DXX-Rebirth: OpenGL: disabling automatic GL sync since VSync is turned off
18:23:54.218087 DXX-Rebirth: OpenGL: using no explicit GPU synchronization
18:23:54.230883 Initializing palette system...
18:23:54.231173 Initializing font system...
```
And then application just closes.

**Comment by vLKp** [comment]
@mmetalfan wrote:
> I would also like to add that I am having similar issues starting the game except I get a general segmentation fault.

This is a different issue.  The original reporter encountered a data use error that caused the program to deliberately abort before something worse happened.  You encountered a crash where the operating system detected a problem.  Additionally, you are using a _very_ outdated version of Rebirth, whereas the original reporter is using a recent snapshot build.

With the information you have provided, the only help I can offer is to suggest that you use a more recent build.  Much has changed since 0.58.1 was tagged.  Any fix for your build would almost certainly not apply to the current code, and any fix in the current code for your problem would not apply cleanly to 0.58.1.

@hfc2x wrote:
> How do I provide that?

Though you are not on Arch Linux, the Arch Linux Wiki has a good and likely usable primer at: https://wiki.archlinux.org/title/Debugging/Getting_traces#Getting_the_trace

In short:

- `gdb -ex r --args ./d1x-rebirth -hogdir .`
- Wait for the game to crash.
- Run `bt full`.  Post the output.

**Comment by mmetalfan** [comment]
Thanks for the pointers. I will try a more recent build. I got this from installing through Lutris so script must be out of date.

**Comment by hfc2x** [comment]
> In short:
> * `gdb -ex r --args ./d1x-rebirth -hogdir .`
> * Wait for the game to crash.
> * Run `bt full`. Post the output.

Here's the terminal output:

```
gdb -ex r --args ./d1x-rebirth -hogdir .                          ✔  40s  
GNU gdb (GDB) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./d1x-rebirth...
Starting program: /home/redacted/Press F/Steam/steamapps/common/Descent/descent/d1x-rebirth -hogdir .
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Type '/home/redacted/Press F/Steam/steamapps/common/Descent/descent/d1x-rebirth -help' for a list of command-line options.

D1X-Rebirth v0.61 0.60.0-beta2-1862-g1c8ff1ac0f13  Aug 10 2022 13:16:51
This is a MODIFIED version of Descent, based on Registered v1.5 Jan 5, 1996.
Copyright (C) 1994, 1995 Parallax Software Corporation
DESCENT is a trademark of Interplay Productions, Inc.
Copyright (C) 2005-2013 Christian Beckhaeuser, 2013-2017 Kp
Using SDL_mixer library v2.6.1
[New Thread 0x7fffe62936c0 (LWP 30796)]
[New Thread 0x7fffe62526c0 (LWP 30797)]
[New Thread 0x7fffe5a046c0 (LWP 30799)]
sdl-joystick: 0 joysticks
[New Thread 0x7fffe4a306c0 (LWP 30800)]
[New Thread 0x7fffcc9036c0 (LWP 30801)]
[New Thread 0x7fffc0f436c0 (LWP 30802)]
[New Thread 0x7fffbbfff6c0 (LWP 30803)]
[New Thread 0x7fffbb7fe6c0 (LWP 30804)]
[New Thread 0x7fffbaffd6c0 (LWP 30805)]
[New Thread 0x7fffba7fc6c0 (LWP 30806)]
[New Thread 0x7fffb9ffb6c0 (LWP 30807)]
[Thread 0x7fffb9ffb6c0 (LWP 30807) exited]
[New Thread 0x7fffb9ffb6c0 (LWP 30808)]
DXX-Rebirth: OpenGL: disabling automatic GL sync since VSync is turned off
double free or corruption (fasttop)

Thread 1 "d1x-rebirth" received signal SIGABRT, Aborted.
0x00007ffff778a4dc in ?? () from /usr/lib/libc.so.6
(gdb) bt full
#0  0x00007ffff778a4dc in  () at /usr/lib/libc.so.6
#1  0x00007ffff773a998 in raise () at /usr/lib/libc.so.6
#2  0x00007ffff772453d in abort () at /usr/lib/libc.so.6
#3  0x00007ffff777e67e in  () at /usr/lib/libc.so.6
#4  0x00007ffff779426c in  () at /usr/lib/libc.so.6
#5  0x00007ffff7796042 in  () at /usr/lib/libc.so.6
#6  0x00007ffff7798a33 in free () at /usr/lib/libc.so.6
#7  0x00007ffff7d47e33 in  () at /usr/lib/libSDL2-2.0.so.0
#8  0x0000555555633397 in dcx::(anonymous namespace)::load_mus_data
    (hook_finished_track=0x555555632f20 <dcx::mix_free_music()>, loop=1, size=<optimized out>, data=<optimized out>)
    at common/arch/sdl/digi_mixer_music.cpp:270
        rw = <optimized out>
        v = @0x7fffffffc8f8: std::vector of length 0, capacity 0
        hoe = @0x7fffffffc8f0: dcx::hmp_open_error::None
        fptr = <optimized out>

                  full_path = {_M_elems = "0\311\377\377\377\177\000\000\000\251mUUU\000\000\004\000\000\000\000\000\000\000\035\247aUUU\000\000font3-1.fnt\000\000U\000\000palette.256\000\263\000\000\000\264\000\000\000\265\000\000\000\266\000\000\000\267\000\000\000\270\000\000\000\271\000\000\000\272\000\000\000\273\000\000\000\274\000\000\000\275\000\000\000\276\000\000\000\277\000\000\000\300\000\000\000\301\000\000\000\302\000\000\000\303\000\000\000\304\000\000\000\305\000\000\000\306\000\000\000\307\000\000\000\310\000\000\000\311\000\000\000\312\000\000\000\313\000\000\000@\256|\316\377\177\000\000p[}\316\377\177\000\000\000\267~\316\377\177\000\000\060\317~\316\377\177\000\000\020\317~\316\377\177\000\000"...}
        bufsize = 0
        hook_finished_track = 0x555555632f20 <dcx::mix_free_music()>
#9  dcx::mix_play_file(char const*, int, void (*)())
    (filename=0x555556165e50 "descent.hmp", loop=loop@entry=1, entry_hook_finished_track=entry_hook_finished_track@entry=0x0) at common/arch/sdl/digi_mixer_music.cpp:157
        v = @0x7fffffffc8f8: std::vector of length 0, capacity 0
        hoe = @0x7fffffffc8f0: dcx::hmp_open_error::None
        fptr = <optimized out>

                  full_path = {_M_elems = "0\311\377\377\377\177\000\000\000\251mUUU\000\000\004\000\000\000\000\000\000\000\035\247aUUU\000\000font3-1.fnt\000\000U\000\000palette.256\000\263\000\000\000\264\000\000\000\265\000\000\000\266\000\000\000\267\000\000\000\270\000\000\000\271\000\000\000\272\000\000\000\273\000\000\000\274\000\000\000\275\000\000\000\276\000\000\000\277\000\000\000\300\000\000\000\301\000\000\000\302\000\000\000\303\000\000\000\304\000\000\000\305\000\000\000\306\000\000\000\307\000\000\000\310\000\000\000\311\000\000\000\312\000\000\000\313\000\000\000@\256|\316\377\177\000\000p[}\316\377\177\000\000\000\267~--Type <RET> for more, q to quit, c to continue without paging--
\316\377\177\000\000\060\317~\316\377\177\000\000\020\317~\316\377\177\000\000"...}
        bufsize = 0
        hook_finished_track = 0x555555632f20 <dcx::mix_free_music()>
#10 0x00005555556037bf in songs_play_file(char const*, int, void (*)())
    (filename=<optimized out>, repeat=repeat@entry=1, hook_finished_track=hook_finished_track@entry=0x0)
    at similar/main/songs.cpp:468
        fptr = <optimized out>
#11 0x0000555555603daf in d1x::songs_play_song(int, int) (songnum=songnum@entry=0, repeat=repeat@entry=1)
    at similar/main/songs.cpp:503
        play = <optimized out>
#12 0x000055555560dadd in d1x::show_titles() () at similar/main/titles.cpp:224
        resolution_at_least_640_480 = <optimized out>
        logo_hires_pcx = <optimized out>
        descent_hires_pcx = <optimized out>
#13 0x000055555557035b in d1x::main
    (argv_gd$b32$GAXDMMBOGAWWEZLUMEZC2MJYGYZC2ZZRMM4GMZRRMFRTAZRRGMzzzzzz=<optimized out>, argc_gc$1c8ff1ac0f137aef68a50611c6714cd1cdf0a043=<optimized out>) at similar/main/inferno.cpp:642
        descent_hog = std::unique_ptr<std::array<char, 4096>> = {get() = 0x555555fb00e0}
        arch_atexit_result = @0x7fffffffd9b5: {<No data fields>}
#14 main(int, char**)
    (argc_gc$1c8ff1ac0f137aef68a50611c6714cd1cdf0a043=<optimized out>, argv_gd$b32$GAXDMMBOGAWWEZLUMEZC2MJYGYZC2ZZRMM4GMZRRMFRTAZRRGMzzzzzz=<optimized out>) at similar/main/inferno.cpp:784
```

**Comment by vLKp** [comment]
I do not get a crash with music configured, but this is at least a start on finding the problem.

**Comment by vLKp** [comment]
[Fix SDL2 double-free on failure to load music](/dxx-rebirth/dxx-rebirth/commit/53d81e78de7606bb95f1896f529533b244238bf5) should fix this.  The problem occurs only in SDL2, and only if the music fails to load.  If the music loads successfully, no double-free occurs, which makes this somewhat harder to reproduce.

Please sync and retest.  If you still see problems, reopen.

**Comment by hfc2x** [comment]
I tested it now, and the game launches. However, there's no music, and terminal output mentions `Music descent.hmp could not be loaded: ModPlug_Load failed`

**Comment by vLKp** [comment]
Yes, your crash was caused because your music fails to load.  If music had been disabled or had loaded successfully, no crash would have occurred even in the bad version.  Rebirth delegates music playing to an external library.  To solve your new problem, you need to determine what library is providing `ModPlug_Load` (a quick search suggests this could be libmikmod or libmodplug), and determine why it cannot use the game's default music file.  Alternatively, you could use a non-default music source, which may work with your existing library.  I've never seen that particular failure mode, so I cannot say why your library is not accepting the music file.
